File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/RingtoneFragment.java
/**
 * @file
 * @brief
 * @author vitamin.caig@gmail.com
 */

package app.zxtune.ui;

import android.app.Activity;
import android.app.Dialog;
import android.net.Uri;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.DialogFragment;
import androidx.fragment.app.FragmentActivity;

import java.util.concurrent.TimeUnit;

import app.zxtune.Permission;
import app.zxtune.R;
import app.zxtune.RingtoneService;
import app.zxtune.TimeStamp;

public class RingtoneFragment extends DialogFragment {

  @Nullable
  static DialogFragment createInstance(FragmentActivity activity, Uri location) {
    if (!Permission.requestSystemSettings(activity)) {
      return null;
    }
    final Bundle args = new Bundle();
    args.putParcelable("url", location);
    final DialogFragment result = new RingtoneFragment();
    result.setArguments(args);
    return result;
  }

  @Override
  public Dialog onCreateDialog(@Nullable Bundle savedInstanceState) {
    final Activity ctx = getActivity();

    final View view = ctx.getLayoutInflater().inflate(R.layout.ringtone, null, false);

    final ViewGroup buttons = view.findViewById(R.id.ringtone_durations);
    fillDurations(buttons);

    return new AlertDialog.Builder(ctx)
                                   .setTitle(R.string.ringtone_create_title)
                                   .setView(view)
                                   .create();
  }

  private void fillDurations(ViewGroup container) {

    final Uri uri = getArguments().getParcelable("url");
    if (uri == null) {
      return;
    }

    final Button.OnClickListener listener = v -> {
      final String tag = (String) v.getTag();
      final int tagValue = Integer.parseInt(tag);
      RingtoneService.execute(getActivity(), uri, TimeStamp.createFrom(tagValue, TimeUnit.SECONDS));
      dismiss();
    };
    for (int idx = 0, lim = container.getChildCount(); idx != lim; ++idx) {
      final Button button = (Button) container.getChildAt(idx);
      button.setOnClickListener(listener);
    }
  }
}

File: apps/zxtune-android/zxtune/src/main/java/app/zxtune/ui/RingtoneFragment.kt
package app.zxtune.ui

import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.provider.Settings
import android.view.View
import android.view.ViewGroup
import androidx.activity.result.contract.ActivityResultContracts
import androidx.annotation.RequiresApi
import androidx.core.view.children
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentActivity
import app.zxtune.R
import app.zxtune.RingtoneService
import app.zxtune.TimeStamp
import app.zxtune.device.Permission
import app.zxtune.ui.utils.FragmentParcelableProperty

class RingtoneFragment : DialogFragment(R.layout.ringtone) {
    companion object {
        fun show(activity: FragmentActivity, uri: Uri) = RingtoneFragment().apply {
            this.uri = uri
        }.show(activity.supportFragmentManager, null)
    }

    private var uri by FragmentParcelableProperty<Uri>()

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        fillUi()
        if (!Permission.canWriteSystemSettings(requireContext())) {
            requestPermission()
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?) =
        super.onCreateDialog(savedInstanceState).apply {
            setTitle(R.string.ringtone_create_title)
        }


    private fun fillUi() {
        val listener = View.OnClickListener { v: View ->
            val tagValue = (v.tag as String).toInt()
            RingtoneService.execute(requireContext(), uri, TimeStamp.fromSeconds(tagValue.toLong()))
            dismiss()
        }
        requireNotNull(view?.findViewById<ViewGroup>(R.id.ringtone_durations)).children.forEach {
            it.setOnClickListener(listener)
        }
    }

    @RequiresApi(23)
    private fun requestPermission() =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            if (!Permission.canWriteSystemSettings(requireContext())) {
                dismiss()
            }
        }.launch(
            Intent(
                Settings.ACTION_MANAGE_WRITE_SETTINGS,
                Uri.parse("package:${requireContext().packageName}")
            )
        )
}
